Local controller - character movement acceleration/deceleration

SAMPROO opened the pull request:
- Character movement
- Acceleration
- Deceleration
- Side & Backwards speed
- Character skin width to 0

- Code not the cleanest yet
- Needs tuning

SAMPROO committed:
Changed character prefab name and changed character controller skin width to 0

SAMPROO committed:
Acceleration and deceleration for keyboard

SAMPROO committed:
Fixed deceleration bug where deceleration would put currentspeed to maxspeed when currentspeed < maxspeed

SAMPROO committed:
Joystick movement seperated from keyboard movement

- joystick movement not yet functional with acceleration and deceleration

SAMPROO committed:
Acceleration and deceleration with keyboard works

- gamepad acceleration and deceleration do not work

SAMPROO committed:
Runnung acceleration with keyboard fixed, code cleanup

SAMPROO committed:
Side and backwards multiplier for speed on keyboard

SAMPROO committed:
Side and backwards run multiplier implemented

SAMPROO committed:
Backwards speed bug fix

SAMPROO committed:
Acceleration and deceleration using physics.....

SAMPROO committed:
Gamepad acceleration and deceleration implemented

oelnenimmat commented on MorkoUnityProject/Assets/Scripts/LocalController.cs:
```diff
@@ -90,33 +151,177 @@ public AvatarPackage Update()
+				oldDirectionGamepad = moveDirectionGamepad;
+			
+			moveDirectionKeyboard = currentMovementSpeed > 0 ? oldDirectionKeyboard.normalized : Vector3.zero;
+			moveDirectionGamepad = currentMovementSpeed > 0 ? oldDirectionGamepad.normalized : Vector3.zero;
```
At this point, use only one variable moveDirection, no reason to differentiate keyboard and gamepad anymore. This removes unnecessary branching below

oelnenimmat commented on MorkoUnityProject/Assets/Scripts/LocalController.cs:
```diff
@@ -90,33 +151,177 @@ public AvatarPackage Update()
-			// Move
-			character.characterController.Move(moveDirection * Time.deltaTime);
 			
+			CalculateMovementSpeed();
```
This tells us absoutely nothing about what variables are changed. Change it so that it returns the movement speed instead of secretly modifying it. See comment on function definition too.

oelnenimmat commented on MorkoUnityProject/Assets/Scripts/LocalController.cs:
```diff
@@ -13,31 +13,66 @@ public class LocalController
+		private float accelerationRun;
+		private float decelerationWalk;
+		private float decelerationRun;
 		private bool isMorko = false;
```
Change player setting only contain one set of values, etiher player or mörkö. 

```
PlayerSettings normalSettings;
PlayerSettings morkoSettings;
PlayerSettings currentSettings => isMorko ? morkoSettings : normalSettings;
```

Change 'walkSpeed' and others to properties e.g. like this
```
private float walkSpeed => currentSettings.walkSpeed;
```

This way, if we change settings, there are fewer places in code we need to change too.

oelnenimmat commented on MorkoUnityProject/Assets/Scripts/LocalController.cs:
```diff
@@ -13,31 +13,66 @@ public class LocalController
+		private Vector3 lastPosition;
+		private Vector3 oldDirectionKeyboard = Vector3.zero;
+		private Vector3 oldDirectionGamepad = Vector3.zero;
+		private bool ran = false;
```
What is this used for? Can we not  infer this value from currentMovementSpeed? If it is for animation, we will still probably use movement speed only. Remove if unnecessary.

oelnenimmat commented on MorkoUnityProject/Assets/Scripts/PlayerSettings.cs:
```diff
@@ -1,19 +1,63 @@
-﻿using UnityEngine;
+﻿using DefaultNamespace;
```
Seems like jetbrains nonsense, remove

oelnenimmat commented on MorkoUnityProject/Assets/Scripts/PlayerSettings.cs:
```diff
@@ -1,19 +1,63 @@
+    public float morkoAccelerationWalk => morkoWalkSpeed / morkoAccelerationTime;
+    public float morkoAccelerationRun => morkoWalkSpeed / morkoAccelerationRunTime;
+    public float morkoDecelerationWalk => morkoWalkSpeed / morkoDecelerationTime;
+    public float morkoDecelerationRun => morkoWalkSpeed / morkoDecelerationRunTime;
```
Take differentiated (morko /normal) settings out of this, and instead use two instances, where one corresponds to morko and one to normal

SAMPROO committed:
PR Fixes

- Keyboard and gamepad is now one
- normal settings and morko settings for playersettings

oelnenimmat commented on MorkoUnityProject/Assets/Scripts/LocalController.cs:
```diff
@@ -158,38 +108,69 @@ public AvatarPackage Update()
 			
-			CalculateMovementSpeed();
+			// Hardcoded min magnitude value where joystick is maxed out
+			bool joystickMaxed = moveDirection.magnitude >= 0.8;
```
This comment can be removed if you make constant like
```
const float joystickMaxThreshold = 0.8f;
```

oelnenimmat commented on MorkoUnityProject/Assets/Scripts/LocalController.cs:
```diff
@@ -158,38 +108,69 @@ public AvatarPackage Update()
+			// Hardcoded min magnitude value where joystick is maxed out
+			bool joystickMaxed = moveDirection.magnitude >= 0.8;
+			
+			// Sneak
```
This comment (and ones below) can be removed:
```
bool sneak = /* Sneak condition*/ ;
bool accelerateWalk = /* accelerate walk condition */;
if (sneak) 
{
    // Sneak code
}
else if (accelerateWalk)
{
    // Code...
}
```
Less comments and more well named variables is better since compiler will check variables on change, but not comments, and some context could be lost when code is changed

oelnenimmat reviewed (changes requested):
Some variables needs to be added to movement calculation section

SAMPROO committed:
PR Code cleanup

SAMPROO committed:
Merge branch 'master' into LocalController
